src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateFont(const FontInfoDesired& fontInfoDesired, _Out_ FontInfo& fontInfo) noexcept
+{
+    return UpdateFont(fontInfoDesired, fontInfo, {}, {});
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern, const SIZE cellSize, const size_t centeringHint) noexcept
+{
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateDpi(const int dpi) noexcept
+{
+    u16 newDPI;
+    RETURN_IF_FAILED(api_narrow(dpi, newDPI));
+
+    if (_api.dpi != newDPI)
+    {
+        _api.dpi = newDPI;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Font);
+    }
+
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateViewport(const SMALL_RECT srNewViewport) noexcept
+{
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::GetProposedFont(const FontInfoDesired& fontInfoDesired, _Out_ FontInfo& fontInfo, const int dpi) noexcept
+try
+{
+    // One day I'm going to implement GDI for AtlasEngine...
+    // Until then this code is work in progress.
+#if 0
+    wil::unique_hfont hfont;
+
+    // This block of code (for GDI fonts) is unfinished.
+    if (fontInfoDesired.IsDefaultRasterFont())
+    {
+        hfont.reset(static_cast<HFONT>(GetStockObject(OEM_FIXED_FONT)));
+        RETURN_HR_IF(E_FAIL, !hfont);
+    }
+    else if (requestedFaceName == DEFAULT_RASTER_FONT_FACENAME)
+    {
+        // GDI Windows Font Mapping reference:
+        // https://msdn.microsoft.com/en-us/library/ms969909.aspx
+
+        LOGFONTW lf;
+        lf.lfHeight = -MulDiv(requestedSize.Y, dpi, 72);
+        lf.lfWidth = 0;
+        lf.lfEscapement = 0;
+        lf.lfOrientation = 0;
+        lf.lfWeight = requestedWeight;
+        lf.lfItalic = FALSE;
+        lf.lfUnderline = FALSE;
+        lf.lfStrikeOut = FALSE;
+        lf.lfCharSet = OEM_CHARSET;
+        lf.lfOutPrecision = OUT_RASTER_PRECIS;
+        lf.lfClipPrecision = CLIP_DEFAULT_PRECIS;
+        lf.lfQuality = PROOF_QUALITY; // disables scaling for rasterized fonts
+        lf.lfPitchAndFamily = FIXED_PITCH | FF_MODERN;
+        // .size() only includes regular characters, but we also want to copy the trailing \0, so +1 it is.
+        memcpy(&lf.lfFaceName[0], &DEFAULT_RASTER_FONT_FACENAME[0], sizeof(DEFAULT_RASTER_FONT_FACENAME));
+
+        hfont.reset(CreateFontIndirectW(&lf));
+        RETURN_HR_IF(E_FAIL, !hfont);
+    }
+
+    if (hfont)
+    {
+// wil::unique_any_t's constructor says: "should not be WI_NOEXCEPT (may forward to a throwing constructor)".
+// The constructor we use by default doesn't throw.
+#pragma warning(suppress : 26447) // The function is declared 'noexcept' but calls function '...' which may throw exceptions (f.6).
+        wil::unique_hdc hdc{ CreateCompatibleDC(nullptr) };
+        RETURN_HR_IF(E_FAIL, !hdc);
+
+        DeleteObject(SelectObject(hdc.get(), hfont.get()));
+
+        SIZE sz;
+        RETURN_HR_IF(E_FAIL, !GetTextExtentPoint32W(hdc.get(), L"M", 1, &sz));

```

Comment:
>L"M" [](http://example.com/codeflow?start=63&length=4)

Now that you experimented with 0, M, space, and presumably other things... perhaps comment why this is M to save the next person

Reply:
Actually I didn't... I only copied DxEngine, but I'm gonna assume because it's usually the widest character in a proportional font for ASCII...

Reply:
I mean it is called an "em" (https://en.wikipedia.org/wiki/Em_(typography)) so I guess we're fine.

Reply:
Wat. Really? Alright I'll document that. 😄